rs/orchestrator/src/upgrade.rs

```diff
@@ -136,20 +140,70 @@ impl Upgrade {
+                    (subnet_id, maybe_proto, maybe_cup)
+                }
+                (None, Some(proto)) => {
+                    // We found a local CUP proto that we can't deserialize. This may only happen
+                    // if this is the first CUP we are reading on a new replica version after an
+                    // upgrade. This means we have to be an assigned node, otherwise we would have
+                    // left the subnet and deleted the CUP before upgrading to this version.
+                    // The only way to leave this branch is via subnet recovery.
+                    self.metrics.critical_error_cup_deserialization_failed.inc();
+
+                    // Try to find the subnet ID by deserializing only the NiDkgId. If it fails
+                    // we will have to recover using failover nodes.
+                    let nidkg_id: NiDkgId = try_from_option_field(proto.signer.clone(), "NiDkgId")
+                        .map_err(|err| {
+                            OrchestratorError::UpgradeError(format!(
+                                "Couldn't deserialize NiDkgId to determine the subnet id: {:?}",
+                                err
+                            ))
+                        })?;
+
+                    let subnet_id = match nidkg_id.target_subnet {
+                        NiDkgTargetSubnet::Local => nidkg_id.dealer_subnet,
+                        NiDkgTargetSubnet::Remote(_) => {
+                            // If this CUP was created by a remote subnet, then it is a genesis/recovery
+                            // CUP. This is the only case in the branch where we can trust the subnet ID
+                            // of the latest registry version, as switching to a registry CUP "resets" the
+                            // "oldest registry version in use" which is responsible for subnet membership.
+                            match self.registry.get_subnet_id(latest_registry_version) {
+                                Ok(subnet_id) => subnet_id,
+                                Err(OrchestratorError::NodeUnassignedError(_, _)) => {
+                                    // If the registry says that we are unassigned, this unassignment
+                                    // must have happened after the registry CUP triggering the upgrade.
+                                    // Otherwise we would have left the subnet before upgrading. This means
+                                    // we will trust the registry and go ahead with removing the node's state
+                                    // including the broken local CUP.
+                                    self.remove_state()?;
+                                    return Ok(None);
+                                }
+                                Err(other) => return Err(other),
+                            }
+                        }
+                    };
+                    (subnet_id, maybe_proto, None)
```

Comment:
One final question: before this change we knew that when we call `get_latest_cup` below that the local cup can be deserialized and consequently that we will be able to obtain `local_cup_height`. If I understand correctly, in this branch we could end up in a situation where we are able to deserialize the NiDKG but not the height from the local CUP proto and consequently fail to obtain `local_cup_height` in `get_latest_cup` below. Do we need a check that ensures that we can obtain the height from the CUP proto and error out otherwise?

Reply:
If we can't determine the height of the local CUP in `get_latest_cup`, then we will fall back to the latest registry CUP before the local CUP. Since we don't have the state of that registry CUP anymore, we still wouldn't make progress, but wait for a higher registry CUP with the correct state.

If the old registry CUP triggers a downgrade, then we should only do the downgrade if the replica version wasn't already retired.

I guess the alternative would indeed be to return an error and accept that we will have to recover on failover nodes in this case.

Reply:
I have a slight tendency towards the latter option -- if we can't even obtain the height we can't be sure whether using the registry CUP is the right decision, e.g., it would seem problematic to me if we attempt to downgrade in such a situation based on the registry CUP even in a case where the replica version is not yet retired.

But maybe I'm missing something. What is your take? Would also be interesting to see what @chmllr thinks.

Reply:
I think that makes sense. To get into this situation we must have already messed up a lot, so maybe aborting earlier is a good idea